From a3d42f72a4c15a231d24839066810b4bda72906a Mon Sep 17 00:00:00 2001
From: Michael Niedermayer <michael@niedermayer.cc>
Date: Tue, 20 Oct 2026 15:10:36 +0200
Subject: [PATCH] avcodec/subviewerdec: do not read the event beyond the packet

Fixes: a3ewdvDE2QSe
Fixes: out of array read
Out of array read Replicated through API with ASAN
Found during triage of the review of https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/24659

Found-by: Marton Balint <cus@passwd.hu>
(cherry picked from commit a3d42f72a4c15a231d24839066810b4bda72906a)
---
 libavcodec/subviewerdec.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/libavcodec/subviewerdec.c b/libavcodec/subviewerdec.c
index 209f69505ff..4a0c8563ac2 100644
--- a/libavcodec/subviewerdec.c
+++ b/libavcodec/subviewerdec.c
@@ -27,6 +27,7 @@
 #include "avcodec.h"
 #include "ass.h"
 #include "codec_internal.h"
+#include "decode.h"
 #include "libavutil/bprint.h"
 
 static int subviewer_event_to_ass(AVBPrint *buf, const char *p)
@@ -51,10 +52,13 @@ static int subviewer_decode_frame(AVCodecContext *avctx, AVSubtitle *sub,
                                   int *got_sub_ptr, const AVPacket *avpkt)
 {
     int ret = 0;
-    const char *ptr = avpkt->data;
+    const char *ptr = ff_decode_subtitle_text(avctx, avpkt);
     FFASSDecoderContext *s = avctx->priv_data;
     AVBPrint buf;
 
+    if (avpkt->size > 0 && !ptr)
+        return AVERROR(ENOMEM);
+
     av_bprint_init(&buf, 0, AV_BPRINT_SIZE_UNLIMITED);
     if (ptr && avpkt->size > 0 && !subviewer_event_to_ass(&buf, ptr))
         ret = ff_ass_add_rect(sub, buf.str, s->readorder++, 0, NULL, NULL);
-- 
2.53.0

